controller/tests/server/update/outcome.ml

```diff
@@ -25,20 +25,21 @@ let slot_spec_to_outcome ({booted_slot; primary_slot; input_versions} : Helpers.
 (* Checks if the state returned by UpdateService matches
    the expected outcome as determined by [slot_spec_to_outcome] *)
 let state_matches_expected_outcome state outcome =
-    match (outcome, state) with
-        | (InstallVsn v1,             Update.Downloading v2) ->
-                (Semver.to_string v1) = v2
-        | (InstallVsn _,              _) ->                         false
-        | (DoNothingOrProduceWarning, Update.ErrorGettingVersionInfo _) -> true
-        | (DoNothingOrProduceWarning, Update.UpToDate _) ->                true
-        | (DoNothingOrProduceWarning, Update.OutOfDateVersionSelected) ->  true
-        | (DoNothingOrProduceWarning, Update.RebootRequired) ->            true
-        | (DoNothingOrProduceWarning, Update.ReinstallRequired) ->         true
-        (* should not _directly_ return to GettingVersionInfo state *)
-        | (DoNothingOrProduceWarning, Update.GettingVersionInfo) ->        false
-        (* all the other states are part of the installation process
-           and are treated as errors *)
-        | (DoNothingOrProduceWarning, _) ->                         false
+    let open Update in
+    match (outcome, state.system_status, state.process_state) with
+    | (InstallVsn v1, NeedsUpdate, Downloading v2) ->
+            (Semver.to_string v1) = v2 &&
+            Option.fold ~none:false ~some:(fun v -> v.latest = v1) state.version_info
+    | (InstallVsn _, _, _) -> false
+    | (DoNothingOrProduceWarning, UpdateError (ErrorGettingVersionInfo _), Sleeping _) -> true
+    | (DoNothingOrProduceWarning, UpToDate, Sleeping _) -> true
+    | (DoNothingOrProduceWarning, OutOfDateVersionSelected, Sleeping _) -> true
+    | (DoNothingOrProduceWarning, RebootRequired, Sleeping _) -> true
+    | (DoNothingOrProduceWarning, ReinstallRequired, Sleeping _) ->true
+    (* should not _directly_ return to GettingVersionInfo state *) (*TODO: redundant now *)
+    | (DoNothingOrProduceWarning, _, Update.GettingVersionInfo) -> false
```

Comment:
It should be merged, yes. It was redundant before and it remains redundant now. I _think_ I added the TODO, because now all the expected process states are made explicit, so it is "even more" redundant. Originally I did add it for documentation purposes, yes.

Reply:
Removed in 1fc88d5